### PR TITLE
More of Player

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/hooker/setuphooks_zh.cpp [hunk 1]
```diff
@@ -2455,7 +2455,6 @@ void Setup_Hooks()
     Hook_Any(0x0049F0A0, TeamPrototype::Heal_All_Objects);
     Hook_Any(0x0049F0E0, TeamPrototype::Iterate_Objects);
     Hook_Any(0x0049F130, TeamPrototype::Count_Team_Instances);
-    Hook_Any(0x0049F160, TeamPrototype::Has_Any_Buildings);
```
Comment (xezon) [inline src/hooker/setuphooks_zh.cpp:2458]: Removal is intentional?

File: src/game/common/staticnamekey.h [hunk 2]
```diff
@@ -135,3 +135,10 @@ extern StaticNameKey g_teamStartsFullKey;
 extern StaticNameKey g_teamReinforcementOriginKey;
 extern StaticNameKey g_teamVeterancyKey;
 extern StaticNameKey g_multiplayerIsLocalKey;
+extern StaticNameKey g_playerIsSkirmishKey;
+extern StaticNameKey g_playerIsPreorderKey;
+extern StaticNameKey g_multiplayerStartIndexKey;
+extern StaticNameKey g_skirmishDifficultyKey;
+extern StaticNameKey g_playerColorKey;
+extern StaticNameKey g_playerNightColorKey;
+extern StaticNameKey g_playerStartMoneyKey;
```
Comment (xezon) [inline src/game/common/staticnamekey.h]: eol

File: src/game/common/rts/player.cpp [hunk 3]
```diff
@@ -1177,3 +1479,456 @@ bool Player::Add_Science(ScienceType t)
...
+    } else if (a == SCIENCE_AVAILABILITY_HIDDEN) {
+        m_hiddenSciences.push_back(t);
+    }
+}
+
+ScienceAvailabilityType Player::Get_Science_Availability_Type_From_String(const Utf8String &availability) const
+{
+    static const char *strings[] = { "Available", "Disabled", "Hidden" };
+
+    for (int i = 0; i < 3; i++) {
```
Comment (xezon) [inline src/game/common/rts/player.cpp:1932]: ARRAY_SIZE(strings)

File: src/game/common/rts/player.cpp [hunk 4]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
 
 void Player::Init_From_Dict(const Dict *d)
```
Comment (xezon) [inline src/game/common/rts/player.cpp:1002]: dict

File: src/game/common/rts/player.cpp [hunk 5]
```diff
@@ -1177,3 +1479,456 @@ bool Player::Add_Science(ScienceType t)
...
+    }
+}
+
+bool Player::Allowed_To_Build(const ThingTemplate *tmplate) const
+{
+    if (!m_canBuildBase && tmplate->Is_KindOf(KINDOF_STRUCTURE)) {
+        return false;
+    }
+
+    return m_canBuildUnits || tmplate->Is_KindOf(KINDOF_STRUCTURE);
```
Comment (xezon) [inline src/game/common/rts/player.cpp:1679]: `return true` ?

File: src/game/common/rts/player.cpp [hunk 6]
```diff
@@ -1177,3 +1479,456 @@ bool Player::Add_Science(ScienceType t)
...
+
+bool Player::Does_Object_Qualify_For_Battle_Plan(Object *obj) const
+{
+    return m_battlePlanBonuses != nullptr && obj != nullptr && obj->Is_Any_KindOf(m_battlePlanBonuses->m_validKindOf)
+        && !obj->Is_Any_KindOf(m_battlePlanBonuses->m_invalidKindOf);
+}
+
+Squad *Player::Get_Hotkey_Squad(int squad)
+{
+    if (squad > 9) {
```
Comment (xezon) [inline src/game/common/rts/player.cpp:1827]: What about < 0 ?

File: src/game/common/rts/player.cpp [hunk 7]
```diff
@@ -1177,3 +1479,456 @@ bool Player::Add_Science(ScienceType t)
...
+            return;
+        }
+    }
+
+    captainslog_dbgassert(false, "removeKindOfProductionCostChange was called with invalid kindOf and percent");
+}
+
+void Player::Set_Attacked_By(int player)
+{
+    captainslog_dbgassert(player >= 0, "Player::setAttackedBy Player index is %d", player);
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: Set_Attacked_By

Comment (xezon) [inline src/game/common/rts/player.cpp]: What about player >= MAX_PLAYER_COUNT ?

File: src/game/common/rts/player.cpp [hunk 8]
```diff
@@ -1177,3 +1479,456 @@ bool Player::Add_Science(ScienceType t)
...
+    m_attackedByPlayer[player] = true;
+    m_lastAttackedByFrame = g_theGameLogic->Get_Frame();
+}
+
+bool Player::Get_Attacked_By(int player) const
+{
+    return m_attackedByPlayer[player];
```
Comment (xezon) [inline src/game/common/rts/player.cpp:1882]: Debug assert missing?

File: src/game/common/rts/player.cpp [hunk 9]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+    m_playerDisplayName = d->Get_UnicodeString(g_playerDisplayNameKey);
+    Utf8String name = d->Get_AsciiString(g_playerNameKey);
+    m_playerName = name;
+    m_playerNameKey = g_theNameKeyGenerator->Name_To_Key(name.Str());
+    bool is_skirmish = false;
+    bool non_skirmish = false;
+    bool exists;
+
+    if (d->Get_Bool(g_playerIsSkirmishKey, &exists)) {
+        for (int i = 0; i < g_theSidesList->Get_Num_Skirmish_Sides(); i++) {
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: side_index

File: src/game/common/rts/player.cpp [hunk 10]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+            Dict &dict = g_theSidesList->Get_Skirmish_Sides_Info(i)->Get_Dict();
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: side_dict

File: src/game/common/rts/player.cpp [hunk 11]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+
+        if (d->Get_Bool(g_playerIsPreorderKey, &exists)) {
+            m_playerIsPreorder = true;
+        }
+
+        if (g_theSidesList->Get_Num_Skirmish_Sides() > 0) {
+            Utf8String civilian("Civilian");
+            bool civilian_found = false;
+            Utf8String str2;
+            int i;
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: side_index

File: src/game/common/rts/player.cpp [hunk 12]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+
+            for (i = 0; i < g_theSidesList->Get_Num_Skirmish_Sides(); i++) {
+                Dict &dict = g_theSidesList->Get_Skirmish_Sides_Info(i)->Get_Dict();
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: side_dict

File: src/game/common/rts/player.cpp [hunk 13]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
-#endif
+    Utf8String faction = d->Get_AsciiString(g_playerFactionKey);
+    PlayerTemplate *pt = g_thePlayerTemplateStore->Find_Player_Template(g_theNameKeyGenerator->Name_To_Key(faction.Str()));
+    captainslog_dbgassert(pt != nullptr,
+        "PlayerTemplate %s not found -- this is an obsolete map (please open and resave in WB)",
+        faction.Str());
+
+    Init(pt);
+    m_playerDisplayName = d->Get_UnicodeString(g_playerDisplayNameKey);
+    Utf8String name = d->Get_AsciiString(g_playerNameKey);
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: name_str

File: src/game/common/rts/player.cpp [hunk 14]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+                        str2.Format("%s%d", dict.Get_AsciiString(g_playerNameKey).Str(), m_multiplayerStartIndex);
+                        civilian_found = true;
+                        break;
+                    }
+                }
+            }
+
+            if (civilian_found) {
+                if (g_theSidesList->Get_Skirmish_Sides_Info(i)->Get_ScriptList() != nullptr) {
+                    Utf8String str1;
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: index_str

File: src/game/common/rts/player.cpp [hunk 15]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+        Set_Player_Type(PLAYER_HUMAN, false);
+
+        if (d->Get_Bool(g_playerIsPreorderKey, &exists)) {
+            m_playerIsPreorder = true;
+        }
+
+        if (g_theSidesList->Get_Num_Skirmish_Sides() > 0) {
+            Utf8String civilian("Civilian");
+            bool civilian_found = false;
+            Utf8String str2;
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: name_index_str

File: src/game/common/rts/player.cpp [hunk 16]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+        Set_Player_Type(PLAYER_COMPUTER, is_skirmish);
+    }
+
+    m_multiplayerStartIndex = d->Get_Int(g_multiplayerStartIndexKey, &exists);
+
+    if (is_skirmish) {
+        Utf8String side = Get_Side();
+        bool side_found = false;
+        Utf8String str2;
+        int i;
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: side_index

File: src/game/common/rts/player.cpp [hunk 17]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+    } else {
+        Set_Player_Type(PLAYER_COMPUTER, is_skirmish);
+    }
+
+    m_multiplayerStartIndex = d->Get_Int(g_multiplayerStartIndexKey, &exists);
+
+    if (is_skirmish) {
+        Utf8String side = Get_Side();
+        bool side_found = false;
+        Utf8String str2;
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: name_index_str

File: src/game/common/rts/player.cpp [hunk 18]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+        }
+
+        if (m_ai != nullptr) {
+            m_ai->Set_Difficulty(difficulty);
+        }
+
+        if (side_found) {
+            m_playerName = str2;
+            Utf8String str1;
+            str1.Format("%d", m_multiplayerStartIndex);
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: index_str

File: src/game/common/rts/player.cpp [hunk 19]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+            ScriptList *list =
+                g_theSidesList->Get_Skirmish_Sides_Info(i)->Get_ScriptList()->Duplicate_And_Qualify(str1, str2, name);
+
+            if (g_theSidesList->Get_Sides_Info(Get_Player_Index())->Get_ScriptList() != nullptr) {
+                g_theSidesList->Get_Sides_Info(Get_Player_Index())->Get_ScriptList()->Delete_Instance();
+            }
+
+            g_theSidesList->Get_Sides_Info(Get_Player_Index())->Set_ScriptList(list);
+
+            for (int j = 0; j < g_theSidesList->Get_Num_Teams(); j++) {
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: team_index

File: src/game/common/rts/player.cpp [hunk 20]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+                Dict *dict = g_theSidesList->Get_Team_Info(j)->Get_Dict();
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: team_dict

File: src/game/common/rts/player.cpp [hunk 21]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+
+                if (dict->Get_AsciiString(g_teamOwnerKey) == name) {
+                    g_theSidesList->Remove_Team(j--);
+                }
+            }
+
+            Utf8String player_name = g_theSidesList->Get_Skirmish_Sides_Info(i)->Get_Dict().Get_AsciiString(g_playerNameKey);
+            for (int j = 0; j < g_theSidesList->Get_Num_Skirmish_Teams(); j++) {
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: team_index

File: src/game/common/rts/player.cpp [hunk 22]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+    m_multiplayerStartIndex = d->Get_Int(g_multiplayerStartIndexKey, &exists);
+
+    if (is_skirmish) {
+        Utf8String side = Get_Side();
+        bool side_found = false;
+        Utf8String str2;
+        int i;
+
+        for (i = 0; i < g_theSidesList->Get_Num_Skirmish_Sides(); i++) {
+            Dict &dict = g_theSidesList->Get_Skirmish_Sides_Info(i)->Get_Dict();
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: side_dict

File: src/game/common/rts/player.cpp [hunk 23]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+                    break;
+                }
+            }
+        }
+
+        GameDifficulty difficulty = static_cast<GameDifficulty>(d->Get_Int(g_skirmishDifficultyKey, &exists));
+        GameDifficulty difficulty2 = g_theScriptEngine->Get_Difficulty();
+
+        if (exists) {
+            difficulty2 = difficulty;
```
Comment (xezon) [inline src/game/common/rts/player.cpp:1119]: This looks wrong, because `difficulty2` is never used.

File: src/game/common/rts/player.cpp [hunk 24]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+        }
+
+        if (m_ai != nullptr) {
+            m_ai->Set_Difficulty(difficulty);
+        }
+
+        if (side_found) {
+            m_playerName = str2;
+            Utf8String str1;
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: index_str

File: src/game/common/rts/player.cpp [hunk 25]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+                        script_keys[0] = g_teamOnCreateScriptKey.Key();
+                        script_keys[1] = g_teamOnIdleScriptKey.Key();
+                        script_keys[2] = g_teamOnUnitDestroyedScriptKey.Key();
+                        script_keys[3] = g_teamOnDestroyedScriptKey.Key();
+                        script_keys[4] = g_teamEnemySightedScriptKey.Key();
+                        script_keys[5] = g_teamAllClearScriptKey.Key();
+                        script_keys[6] = g_teamProductionConditionKey.Key();
+                        Utf8String script;
+
+                        for (int key = 0; key < 7; key++) {
```
Comment (xezon) [inline src/game/common/rts/player.cpp:1174]: ARRAY_SIZE(script_keys)

File: src/game/common/rts/player.cpp [hunk 26]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+                            script = new_dict.Get_AsciiString(script_keys[key], &exists);
+
+                            if (exists && !script.Is_Empty()) {
+                                str.Format("%s%d", script.Str(), m_multiplayerStartIndex);
+                                new_dict.Set_AsciiString(script_keys[key], str);
+                            }
+                        }
+
+                        for (int key = 0; key < 16; key++) {
+                            Utf8String script2;
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: script_key_str

File: src/game/common/rts/player.cpp [hunk 27]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+        m_playerNightColor = m_playerColor;
+    }
+
+    color = d->Get_Int(g_playerNightColorKey, &exists);
+
+    if (exists) {
+        m_playerNightColor = color | 0xFF000000;
+    }
+
+    int amount = d->Get_Int(g_playerStartMoneyKey, &exists);
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: money_value

File: src/game/common/rts/player.cpp [hunk 28]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+                        for (int key = 0; key < 7; key++) {
+                            script = new_dict.Get_AsciiString(script_keys[key], &exists);
+
+                            if (exists && !script.Is_Empty()) {
+                                str.Format("%s%d", script.Str(), m_multiplayerStartIndex);
+                                new_dict.Set_AsciiString(script_keys[key], str);
+                            }
+                        }
+
+                        for (int key = 0; key < 16; key++) {
```
Comment (xezon) [inline src/game/common/rts/player.cpp:1183]: MAX_LIST_COUNT perhaps

File: src/game/common/rts/player.cpp [hunk 29]
```diff
@@ -946,9 +1001,256 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+                Dict *dict = g_theSidesList->Get_Team_Info(j)->Get_Dict();
+
+                if (dict->Get_AsciiString(g_teamOwnerKey) == name) {
+                    g_theSidesList->Remove_Team(j--);
+                }
+            }
+
+            Utf8String player_name = g_theSidesList->Get_Skirmish_Sides_Info(i)->Get_Dict().Get_AsciiString(g_playerNameKey);
+            for (int j = 0; j < g_theSidesList->Get_Num_Skirmish_Teams(); j++) {
+                Dict *dict = g_theSidesList->Get_Skirmish_Team_Info(j)->Get_Dict();
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: team_dict

File: src/hooker/setuphooks_zh.cpp (hunk 1 above)
Comment (jonwil) [reply src/hooker/setuphooks_zh.cpp:2458]: Yes.

File: src/game/common/staticnamekey.h (hunk 2 above)
Comment (jonwil) [reply src/game/common/staticnamekey.h]: fixed.

File: src/game/common/rts/player.cpp (hunk 5 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp:1679]: No this is correct.

File: src/game/common/rts/player.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 8 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp:1882]: Nope.

File: src/game/common/rts/player.cpp (hunk 9 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 10 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 11 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 12 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 13 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 14 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 15 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 16 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 17 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 18 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 19 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 20 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 21 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 22 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 23 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp:1119]: fixed.

File: src/game/common/rts/player.cpp (hunk 24 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 26 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 27 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 29 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

**Review by xezon — APPROVED** [review]
Looks good.

**Pull request merged**